trinity/protocol/common/managers.py

```diff
@@ -214,3 +237,129 @@ def _send_sub_proto_request(self, request: TRequest) -> None:
+    #
+    @property
+    def subscription_msg_types(self) -> Set[Type[Command]]:
+        return {self._response_msg_type}
+
+    msg_queue_maxsize = 100
+
+    response_timout: int = 60
+
+    pending_request: Tuple[float, 'asyncio.Future[TMsg]'] = None
+
+    _peer: TPeer
+
+    def __init__(self, peer: TPeer, response_msg_type: Type[Command], token: CancelToken) -> None:
+        super().__init__(token)
+        self._peer = peer
+        self.response_times = ResponseTimeTracker()
+        self._response_msg_type = response_msg_type
+
+    async def message_candidates(
+            self,
+            make_request: Callable[[], None],
+            logger: logging.Logger,
+            timeout: int) -> 'AsyncGenerator[TMsg, None]':
+        """
+        Make a request and iterate through candidates for a valid response.
+
+        To mark a response as valid, use the `complete_if_valid` context manager. Within it,
+        validate the received message. If no ValidationError is raised, then the
+        message will be considered delivered, and the response closed.
+        """
+        self._request(make_request)
+        while self._is_pending():
+            yield await self._get_message(timeout)
+
+    @contextmanager
+    def complete_if_valid(self, item_count: int) -> Generator[None, None, None]:
+        try:
+            yield
+        except ValidationError as err:
+            self.logger.debug(
+                "Response validation failed for pending %s request from peer %s: %s",
+                self.response_msg_name,
+                self._peer,
+                err,
+            )
+        else:
+            self.pending_request = None
+            self.response_times.add(self.last_response_time, item_count)
+
+    @property
+    def response_msg_name(self) -> str:
+        return self._response_msg_type.__name__
+
+    #
+    # Service API
+    #
+    async def _run(self) -> None:
+        self.logger.debug("Launching %s for peer %s", self.__class__.__name__, self._peer)
+
+        with self.subscribe_peer(self._peer):
+            while self.is_running:
+                peer, cmd, msg = await self.wait(self.msg_queue.get())
+                if peer != self._peer:
+                    self.logger.error("Unexpected peer: %s  expected: %s", peer, self._peer)
+                    continue
+                elif isinstance(cmd, self._response_msg_type):
+                    await self._handle_msg(cast(TMsg, msg))
+                else:
+                    self.logger.warning("Unexpected message type: %s", cmd.__class__.__name__)
+
+    async def _handle_msg(self, msg: TMsg) -> None:
+        if self.pending_request is None:
+            self.logger.debug(
+                "Got unexpected %s message from %", self.response_msg_name, self._peer
+            )
+            return
+
+        send_time, future = self.pending_request
+        self.last_response_time = time.perf_counter() - send_time
+        future.set_result(msg)
```

Comment:
Do I understand correctly that rather than only setting the result if we find a match, we instead set the result for every message of the correct type and then later we validate the result that was set, and if it doesn't validate we initialize a new future and start the process over?

Reply:
Yes